polkadot/node/core/pvf/src/artifacts.rs

```diff
@@ -55,18 +55,56 @@
 //!    older by a predefined parameter. This process is run very rarely (say, once a day). Once the
 //!    artifact is expired it is removed from disk eagerly atomically.
 
-use crate::host::PrepareResultSender;
+use crate::{host::PrepareResultSender, LOG_TARGET};
 use always_assert::always;
+use polkadot_core_primitives::Hash;
 use polkadot_node_core_pvf_common::{error::PrepareError, prepare::PrepareStats, pvf::PvfPrepData};
 use polkadot_node_primitives::NODE_VERSION;
 use polkadot_parachain_primitives::primitives::ValidationCodeHash;
 use polkadot_primitives::ExecutorParamsHash;
 use std::{
 	collections::HashMap,
 	path::{Path, PathBuf},
+	str::FromStr as _,
 	time::{Duration, SystemTime},
 };
 
+macro_rules! concat_const {
+	($x:expr, $y:expr) => {{
+		// ensure inputs to be strings
+		const _: &str = $x;
+		const _: &str = $y;
+
+		const X: &[u8] = $x.as_bytes();
+		const Y: &[u8] = $y.as_bytes();
+		const XL: usize = X.len();
+		const YL: usize = Y.len();
+		const L: usize = XL + YL;
+
+		const fn concat() -> [u8; L] {
+			let mut cat = [0u8; L];
+			let mut i = 0;
+			while i < XL {
+				cat[i] = X[i];
+				i += 1;
+			}
+			while i < L {
+				cat[i] = Y[i - XL];
+				i += 1;
+			}
+			cat
+		}
+
+		// SAFETY: safe because x and y are ensured to be valid
```

Comment:
I'm not so confident that this is always safe because unicode has a lot of edge cases. Both inputs are unicode, and I'm not 100% sure it's sound to concatenate unicode bytes, even if separately they are valid unicode strings. I asked AI:

> Concatenating the bytes of two Unicode strings is not a safe operation because Unicode strings can contain multi-byte characters. If you concatenate two strings at a byte level, you risk splitting a multi-byte character, which can lead to incorrect and unpredictable results.

Of course, the way it's used right now is fine because we only have ASCII inputs, but there's a chance of this macro being mis-used in the future. I would just go with a non-const solution, e.g. use `format!` inside a `OnceCell` or something. There's not really a need to complicate things with a macro.

Reply:
Yeah, the unsafe is unnessary here.

Reply:
I was wrong. Tried a couple of way, couldn't get rid of the unsafe.. AFAIK, `&str` has the same layout as `&[u8]`, so this should be safe in Rust.

> There's not really a need to complicate things with a macro.

I tried a couple of ways. This is by far the simplest from usage perspective, as it is indeed a constant. But if you are really uncomfortable about this, I could replace it with a function. Your call. @mrcnski